planner/core/find_best_task.go

```diff
@@ -849,7 +863,110 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 }
 
-func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *property.StatsInfo) {
+// pushDownSelAndResolveVirtualCols push some filters down to coprocessor and resolve virtual columns.
+//	1. push down filters and check if there is virtual column.
+//	2. substitute virtual columns in TableScan's Schema to corresponding physical columns.
+//	3. substitute virtual columns in TableScan's filters and push them down.
+//	4. add a Projection upon this DataSource.
+func (ds *DataSource) pushDownSelAndResolveVirtualCols(copTask *copTask, path *accessPath, stats *property.StatsInfo) (t task, err error) {
+	// step 1
+	t = copTask
+	if copTask.indexPlan != nil {
+		if err := ds.addPushedDownSelection(copTask, path); err != nil {
+			return invalidTask, err
+		}
+	}
+	if copTask.tablePlan == nil { // don't need to handle virtual columns in IndexScan
+		return
+	}
+	ts, ok := copTask.tablePlan.(*PhysicalTableScan)
+	if !ok {
+		return invalidTask, errors.Errorf("type assertion fail, expect PhysicalTableScan, but got %v", reflect.TypeOf(copTask.tablePlan))
+	}
+	if len(ds.virtualColExprs) == 0 {
+		ds.addPushedDownTableScan(copTask, ts, stats)
+		return
+	}
+
+	// step 2
+	ds.substituteVirtualColumns(ts)
+
+	// step 3
+	for i, expr := range ts.filterCondition {
+		ts.filterCondition[i] = expression.ColumnSubstitute(expr, ds.virtualColSchema, ds.virtualColExprs)
+	}
+	var cantBePushed []expression.Expression
+	_, ts.filterCondition, cantBePushed = expression.ExpressionsToPB(ds.ctx.GetSessionVars().StmtCtx, ts.filterCondition, ds.ctx.GetClient())
+	ds.addPushedDownTableScan(copTask, ts, stats)
+	if len(cantBePushed) > 0 { // for filters cannot be pushed down, we add a root Selection to handle them
+		sel := PhysicalSelection{Conditions: cantBePushed}.Init(ds.ctx, stats)
+		t = sel.attach2Task(copTask)
+	}
+
+	// step 4
+	projExprs := make([]expression.Expression, 0, len(ds.Schema().Columns))
+	for _, c := range ds.Schema().Columns {
+		projExprs = append(projExprs, expression.ColumnSubstitute(c, ds.virtualColSchema, ds.virtualColExprs))
+	}
+	proj := PhysicalProjection{Exprs: projExprs}.Init(ds.ctx, stats)
+	proj.SetSchema(ds.Schema())
+	t = proj.attach2Task(t)
+	return
+}
+
+// substituteVirtualColumns substitute virtual columns in TableScan's Schema to physical columns.
+func (ds *DataSource) substituteVirtualColumns(ts *PhysicalTableScan) {
+	// clone a new Schema to modify for safety
+	schema := ts.Schema().Clone()
+	colInfos := make([]*model.ColumnInfo, 0, len(ts.Columns))
+	colInfos = append(colInfos, ts.Columns...)
+
+	// derive physical columns from virtual columns
+	phyCols := make(map[int64]*expression.Column)        // physical columns this ts already has
+	virCols := make(map[int64]*expression.Column)        // virtual columns this ts has
+	phyColsFromVir := make(map[int64]*expression.Column) // physical columns derived from virtual columns
+	for i := 0; i < len(schema.Columns); i++ {
+		if i < len(colInfos) {
```

Comment:
It's better to early break the loop to avoid unnessary iteration. e.g.
```go
for i := 0; i < len(schema.Columns); i++ {
	if i >= len(colInfos) {
		break
	}
}
```